ath79: Fix recognized RAM size for TP-Link Archer C6 V2 (US) and convert to nvmem layout

rodsmar opened the pull request:
This commit fixes a bug that RAM size autodetection not working properly, by cutting half of total memory size, setting statically RAM size in dts file.

Run-tested on my TP-Link Archer C6 V2 (US).

rodsmar committed:
ath79: Fix recognized RAM size for TP-Link Archer C6 V2 (US)

This commit fixes a bug that RAM size autodetection not working properly, by cutting half of total memory size, setting statically RAM size in dts file.

Run-tested on my TP-Link Archer C6 V2 (US).

Signed-off-by: Rodrigo B. de Sousa Martins <rodrigo.sousa.577@gmail.com>

rodsmar commented:
This PR needs to be backported for 23.05 branch.

I can do this, after merge in main branch.

DragonBluep commented on target/linux/ath79/dts/qca9563_tplink_archer-c6-v2-us.dts:
```diff
@@ -6,6 +6,11 @@
 	compatible = "tplink,archer-c6-v2-us", "qca,qca9563";
 	model = "TP-Link Archer C6 v2 (US) / A6 v2 (US/TW)";
 
+	memory {
```
```suggestion
	memory@0 {
```

DragonBluep commented on target/linux/ath79/dts/qca9563_tplink_archer-c6-v2-us.dts:
```diff
@@ -6,6 +6,11 @@
 
+	memory {
+		device_type = "memory";
+		reg = <0x00000000 0x8000000>;
```
```suggestion
		reg = <0x00000000 0x08000000>;
```

DragonBluep reviewed (commented):
What is the `dmesg` output without this patch?

rodsmar commented:
> What is the `dmesg` output without this patch?

Now, kernel has allocated all memory size... So bizarre...

dmesg with 23.05.2 release

`[    0.000000] Linux version 5.15.137 (builder@buildhost) (mips-openwrt-linux-musl-gcc (OpenWrt GCC 12.3.0 r23630-842932a63d) 12.3.0, GNU ld (GNU Binutils) 2.40.0) #0 Tue Nov 14 13:38:11 2023
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019750 (MIPS 74Kc)
[    0.000000] MIPS: machine is TP-Link Archer C6 v2 (US) / A6 v2 (US/TW)
[    0.000000] SoC: Qualcomm Atheros QCA956X ver 1 rev 0
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32480
[    0.000000] Kernel command line: console=ttyS0,115200n8 rootfstype=squashfs,jffs2
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 120820K/131072K available (6079K kernel code, 597K rwdata, 780K rodata, 1240K init, 217K bss, 10252K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 51
[    0.000000] CPU clock: 775.000 MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 4932285024 ns
[    0.000001] sched_clock: 32 bits at 387MHz, resolution 2ns, wraps every 5541893118ns
[    0.008292] Calibrating delay loop... 385.84 BogoMIPS (lpj=1929216)
[    0.074814] pid_max: default: 32768 minimum: 301
[    0.080514] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.088251] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.102738] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.113154] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.120510] pinctrl core: initialized pinctrl subsystem
[    0.127761] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.134566] thermal_sys: Registered thermal governor 'step_wise'
[    0.145807] clocksource: Switched to clocksource MIPS
[    0.158812] NET: Registered PF_INET protocol family
[    0.164182] IP idents hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.172723] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.181615] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.189800] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.197891] TCP bind hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.205335] TCP: Hash tables configured (established 1024 bind 1024)
[    0.212194] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.219148] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.226915] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.232922] PCI: CLS 0 bytes, default 32
[    0.240409] workingset: timestamp_bits=14 max_order=15 bucket_order=1
[    0.251908] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.258137] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.269802] Block layer SCSI generic (bsg) driver version 0.4 loaded (major 251)
[    0.281248] pinctrl-single 1804002c.pinmux: 544 pins, size 68
[    0.288534] Serial: 8250/16550 driver, 16 ports, IRQ sharing enabled
[    0.297781] printk: console [ttyS0] disabled
[    0.302352] 18020000.uart: ttyS0 at MMIO 0x18020000 (irq = 9, base_baud = 1562500) is a 16550A
[    0.311501] printk: console [ttyS0] enabled
[    0.320572] printk: bootconsole [early0] disabled
[    0.348797] spi-nor spi0.0: w25q128 (16384 Kbytes)
[    0.353817] 6 fixed-partitions partitions found on MTD device spi0.0
[    0.360501] OF: Bad cell count for /ahb/spi@1f000000/flash@0/partitions
[    0.367390] OF: Bad cell count for /ahb/spi@1f000000/flash@0/partitions
[    0.374675] OF: Bad cell count for /ahb/spi@1f000000/flash@0/partitions
[    0.381603] OF: Bad cell count for /ahb/spi@1f000000/flash@0/partitions
[    0.388724] Creating 6 MTD partitions on "spi0.0":
[    0.393691] 0x000000000000-0x000000020000 : "factory-boot"
[    0.404439] 0x000000020000-0x000000030000 : "info"
[    0.410528] 0x000000030000-0x000000050000 : "u-boot"
[    0.418288] 0x000000050000-0x000000fd0000 : "firmware"
[    0.424863] 2 uimage-fw partitions found on MTD device firmware
[    0.431039] Creating 2 MTD partitions on "firmware":
[    0.436200] 0x000000000000-0x000000238c83 : "kernel"
[    0.441332] mtd: partition "kernel" doesn't end on an erase/write block -- force read-only
[    0.452354] 0x000000238c83-0x000000f80000 : "rootfs"
[    0.457552] mtd: partition "rootfs" doesn't start on an erase/write block boundary -- force read-only
[    0.468144] mtd: setting mtd5 (rootfs) as root device
[    0.474253] 1 squashfs-split partitions found on MTD device rootfs
[    0.480710] 0x0000006b0000-0x000000f80000 : "rootfs_data"
[    0.487389] 0x000000fd0000-0x000000ff0000 : "tplink"
[    0.494968] 0x000000ff0000-0x000001000000 : "art"
[    1.229917] switch0: Atheros AR8337 rev. 2 switch registered on mdio.0
[    1.887452] ag71xx 19000000.eth: connected to PHY at mdio.0:00 [uid=004dd036, driver=Atheros AR8216/AR8236/AR8316]
[    1.898775] eth0: Atheros AG71xx at 0xb9000000, irq 4, mode: sgmii
[    1.905666] i2c_dev: i2c /dev entries driver
[    1.912807] NET: Registered PF_INET6 protocol family
[    1.929602] Segment Routing with IPv6
[    1.933492] In-situ OAM (IOAM) with IPv6
[    1.937713] NET: Registered PF_PACKET protocol family
[    1.943011] bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
[    1.956709] 8021q: 802.1Q VLAN Support v1.8
[    1.962221] PCI host bridge to bus 0000:00
[    1.966520] pci_bus 0000:00: root bus resource [mem 0x12000000-0x13ffffff]
[    1.973627] pci_bus 0000:00: root bus resource [io  0x0000]
[    1.979392] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    1.987627] pci 0000:00:00.0: [168c:0056] type 00 class 0x028000
[    1.993871] pci 0000:00:00.0: reg 0x10: [mem 0x00000000-0x001fffff 64bit]
[    2.001021] pci 0000:00:00.0: PME# supported from D0 D3hot
[    2.007649] pci_bus 0000:00: busn_res: [bus 00-ff] end is updated to 00
[    2.014516] pci 0000:00:00.0: BAR 0: assigned [mem 0x12000000-0x121fffff 64bit]
[    2.033158] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    2.047370] Freeing unused kernel image (initmem) memory: 1240K
[    2.053496] This architecture does not have kernel memory protection.
[    2.060180] Run /sbin/init as init process
[    2.064407]   with arguments:
[    2.064411]     /sbin/init
[    2.064416]   with environment:
[    2.064420]     HOME=/
[    2.064424]     TERM=linux
[    2.692602] init: Console is alive
[    2.696723] init: - watchdog -
[    4.082103] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    4.144444] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    4.162959] init: - preinit -
[    5.371681] random: jshn: uninitialized urandom read (4 bytes read)
[    5.601731] random: jshn: uninitialized urandom read (4 bytes read)
[    5.763371] random: jshn: uninitialized urandom read (4 bytes read)
[    7.106614] Atheros AR8216/AR8236/AR8316 mdio.0:00: Port 1 is up
[    7.114990] eth0: link up (1000Mbps/Full duplex)
[    7.132261] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[    7.148215] IPv6: ADDRCONF(NETDEV_CHANGE): eth0.1: link becomes ready
[    7.155370] random: procd: uninitialized urandom read (4 bytes read)
[    8.186771] Atheros AR8216/AR8236/AR8316 mdio.0:00: Port 5 is up
[    9.413406] jffs2_scan_eraseblock(): End of filesystem marker found at 0x10000
[    9.420948] jffs2_build_filesystem(): unlocking the mtd device...
[    9.420956] done.
[    9.429321] jffs2_build_filesystem(): erasing all blocks after the end marker...
[   38.837973] done.
[   38.847737] jffs2: notice: (405) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
[   38.865350] mount_root: overlay filesystem has not been fully initialized yet
[   38.881252] mount_root: switching to jffs2 overlay
[   38.889484] overlayfs: upper fs does not support tmpfile.
[   39.425030] urandom-seed: Seed file not found (/etc/urandom.seed)
[   39.484274] eth0: link down
[   39.507639] procd: - early -
[   39.510919] procd: - watchdog -
[   40.193064] procd: - watchdog -
[   40.196758] procd: - ubus -
[   40.262070] random: ubusd: uninitialized urandom read (4 bytes read)
[   40.270373] random: ubusd: uninitialized urandom read (4 bytes read)
[   40.277496] random: ubusd: uninitialized urandom read (4 bytes read)
[   40.291149] procd: - init -
[   41.265997] random: ubus: uninitialized urandom read (4 bytes read)
[   41.330329] random: jsonfilter: uninitialized urandom read (4 bytes read)
[   41.341713] kmodloader: loading kernel modules from /etc/modules.d/*
[   41.867552] GACT probability on
[   41.871951] Mirror/redirect action on
[   41.907119] u32 classifier
[   41.909931]     input device check on
[   41.913711]     Actions configured
[   41.969525] Loading modules backported from Linux version v6.1.24-0-g0102425ac76b
[   41.977321] Backport generated by backports.git v5.15.92-1-44-gd6ea70fafd36
[   42.499412] ath10k 5.15 driver, optimized for CT firmware, probing pci device: 0x56.
[   42.526756] ath10k_pci 0000:00:00.0: enabling device (0000 -> 0002)
[   42.533537] ath10k_pci 0000:00:00.0: pci irq legacy oper_irq_mode 1 irq_mode 0 reset_mode 0
[   43.336196] urngd: v1.0.2 started.
[   45.078839] random: crng init done
[   45.082371] random: 21 urandom warning(s) missed due to ratelimiting
[   46.940233] ath10k_pci 0000:00:00.0: qca9888 hw2.0 target 0x01000000 chip_id 0x00000000 sub 0000:0000
[   46.949817] ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   46.969090] ath10k_pci 0000:00:00.0: firmware ver 10.4b-ct-9888-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 59e741e7
[   47.308179] ath10k_pci 0000:00:00.0: Loading BDF type 0
[   47.321937] ath10k_pci 0000:00:00.0: board_file api 2 bmi_id 0:24 crc32 ae4cbea5
[   49.095124] ath10k_pci 0000:00:00.0: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[   49.102927] ath10k_pci 0000:00:00.0: msdu-desc: 2500  skid: 32
[   49.153053] ath10k_pci 0000:00:00.0: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[   49.164399] ath10k_pci 0000:00:00.0: wmi print 'free: 114572 iram: 12644 sram: 29508'
[   49.400411] ath10k_pci 0000:00:00.0: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-file max-sta 32 raw 0 hwcrypto 1
[   49.509464] ath: EEPROM regdomain: 0x0
[   49.509485] ath: EEPROM indicates default country code should be used
[   49.509490] ath: doing EEPROM country->regdmn map search
[   49.509504] ath: country maps to regdmn code: 0x3a
[   49.509511] ath: Country alpha2 being used: US
[   49.509517] ath: Regpair used: 0x3a
[   49.602668] ath: EEPROM regdomain: 0x0
[   49.602691] ath: EEPROM indicates default country code should be used
[   49.602697] ath: doing EEPROM country->regdmn map search
[   49.602710] ath: country maps to regdmn code: 0x3a
[   49.602717] ath: Country alpha2 being used: US
[   49.602723] ath: Regpair used: 0x3a
[   49.614668] ieee80211 phy1: Selected rate control algorithm 'minstrel_ht'
[   49.616672] ieee80211 phy1: Atheros AR9561 Rev:0 mem=0xb8100000, irq=2
[   49.642368] kmodloader: done loading kernel modules from /etc/modules.d/*
[   64.099176] eth0: link up (1000Mbps/Full duplex)
[   64.103988] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   64.126351] br-iot: port 1(eth0.50) entered blocking state
[   64.132041] br-iot: port 1(eth0.50) entered disabled state
[   64.138033] device eth0.50 entered promiscuous mode
[   64.143087] device eth0 entered promiscuous mode
[   64.192400] br-iot: port 1(eth0.50) entered blocking state
[   64.198145] br-iot: port 1(eth0.50) entered forwarding state
[   64.466889] br-lan: port 1(eth0.20) entered blocking state
[   64.472583] br-lan: port 1(eth0.20) entered disabled state
[   64.524676] device eth0.20 entered promiscuous mode
[   64.566100] br-lan: port 1(eth0.20) entered blocking state
[   64.571789] br-lan: port 1(eth0.20) entered forwarding state
[   64.756751] br-visita: port 1(eth0.60) entered blocking state
[   64.762709] br-visita: port 1(eth0.60) entered disabled state
[   64.785264] device eth0.60 entered promiscuous mode
[   64.809275] br-visita: port 1(eth0.60) entered blocking state
[   64.815233] br-visita: port 1(eth0.60) entered forwarding state
[   65.146130] IPv6: ADDRCONF(NETDEV_CHANGE): br-iot: link becomes ready
[   68.666067] ath: EEPROM regdomain: 0x804c
[   68.666092] ath: EEPROM indicates we should expect a country code
[   68.666099] ath: doing EEPROM country->regdmn map search
[   68.666104] ath: country maps to regdmn code: 0x3b
[   68.666110] ath: Country alpha2 being used: BR
[   68.666116] ath: Regpair used: 0x3b
[   68.666122] ath: regdomain 0x804c dynamically updated by user
[   68.666213] ath: EEPROM regdomain: 0x804c
[   68.666220] ath: EEPROM indicates we should expect a country code
[   68.666225] ath: doing EEPROM country->regdmn map search
[   68.666229] ath: country maps to regdmn code: 0x3b
[   68.666235] ath: Country alpha2 being used: BR
[   68.666240] ath: Regpair used: 0x3b
[   68.666245] ath: regdomain 0x804c dynamically updated by user
[ 1110.334265] Atheros AR8216/AR8236/AR8316 mdio.0:00: Port 2 is up
`

DragonBluep commented:
> Now, kernel has allocated all memory size... So bizarre...

So, I think it's a hardware issue. Perhaps some of the address lines on your device are no longer reliable...

> [ 0.000000] Normal [mem 0x0000000000000000-0x0000000007ffffff]

Yes kernel can get 128MB RAM now.

neheb commented:
since you're touching this device, can you apply"

```
--- a/target/linux/ath79/dts/qca9563_tplink_archer-c6-v2-us.dts
+++ b/target/linux/ath79/dts/qca9563_tplink_archer-c6-v2-us.dts
@@ -129,10 +129,20 @@
 				read-only;
 			};
 
-			art: partition@ff0000 {
+			partition@ff0000 {
 				label = "art";
 				reg = <0xff0000 0x010000>;
 				read-only;
+
+				nvmem-layout {
+					compatible = "fixed-layout";
+					#address-cells = <1>;
+					#size-cells = <1>;
+
+					calibration_art_1000: calibration@1000 {
+						reg = <0x1000 0x440>;
+					};
+				};
 			};
 		};
 	};
@@ -144,8 +154,6 @@
 };
 
 &wmac {
-	mtd-cal-data = <&art 0x1000>;
-
-	nvmem-cells = <&macaddr_info_8>;
-	nvmem-cell-names = "mac-address";
+	nvmem-cells = <&macaddr_info_8>, <&calibration_art_1000>;
+	nvmem-cell-names = "mac-address", "calibration";
 };
```

?

rodsmar commented:
> +	nvmem-cells = <&macaddr_info_8>, <&calibration_art_1000>;
> +	nvmem-cell-names = "mac-address", "calibration";

Totally agree. Pushed your sugested changes.



> > Now, kernel has allocated all memory size... So bizarre...
> 
> So, I think it's a hardware issue. Perhaps some of the address lines on your device are no longer reliable...
> 
> > [ 0.000000] Normal [mem 0x0000000000000000-0x0000000007ffffff]
> 
> Yes kernel can get 128MB RAM now.

Hmm... maybe you're right. Perhaps i'm can't use kmod-ath10k and ath10k-firmware (without Candela patches) stably. It is related?

rodsmar commented:
@neheb 

I'm using this router with Distributed Switch Architecture (DSA) enabled, but maybe i not include in this PR since depends of other PR which is WIP. (#13589 and #13832 )

neheb commented:
Definitely not

hauke commented:
Please slit this into two commits, one fixing the DDR size and one changing the calibration data. Both changes are unrelated.

rodsmar commented:
> Please slit this into two commits, one fixing the DDR size and one changing the calibration data. Both changes are unrelated.

Alright, but i need to open another PR for calibration data?

hauke commented:
You can do both changes in this PR, just use two commits for it.

rodsmar committed:
ath79: Convert TP-Link Archer C6 V2 (US) to NVMEM layout

This commit convert Archer C6 V2, US variant, to NVMEM layout.

Suggested-by: Rosen Penev <rosenp@gmail.com>
Signed-off-by: Rodrigo B. de Sousa Martins <rodrigo.sousa.577@gmail.com>

rodsmar commented:
> You can do both changes in this PR, just use two commits for it.

Done! Sorry for delay.

hauke commented:
Maybe the `detect_memory_region` function https://elixir.bootlin.com/linux/v6.1.76/source/arch/mips/kernel/setup.c#L100 saw the magic already earlier in one of your boot ups. Do you still see the problem?

DragonBluep commented:
When I added the 6.6 testing kernel for the ath79 target, I reviewed the memory detect code. It is indeed a bug. There is a very small probability that we can get the wrong memory size here. I've sent the fix patch to the upstream.
https://patchwork.kernel.org/project/linux-mips/patch/TYAP286MB0315E609C476B86E22700626BC292@TYAP286MB0315.JPNP286.PROD.OUTLOOK.COM/

rodsmar commented:
Great! When this PR is merged, i should drop this? (or we can make a patch into pending-5.15-23.05-/6.1)

DragonBluep commented:
> Great! When this PR is merged, i should drop this?

Yep I think.

> (or we can make a patch into pending-5.15-23.05-/6.1)

You can backport it here if you like. The 6.6 kernel PR should still take several weeks to be accepted.
